### PR TITLE
[release/v2.7] [WIP] Add configmap/secret based file delivery for v2prov

### PR DESCRIPTION
**Pull request by Oats87**
## Issue: 
https://github.com/rancher/rancher/issues/39994
 
## Problem
v2prov does not have a way to allow arbitrary files to be delivered to downstream nodes
 
## Solution
Allow files to be sourced from configmap and/or secret, in a manner very similar to the way volume mounts work for Kubernetes pods.

### CODE REVIEW

File: pkg/provisioningv2/rke2/planner/config.go [hunk 1]
```diff
@@ -413,6 +435,72 @@ func configFile(controlPlane *rkev1.RKEControlPlane, filename string) string {
...
+						for _, v := range fs.Secret.Items {
+							file := plan.File{
+								Path:    v.Path,
+								Content: base64.StdEncoding.EncodeToString(secret.Data[v.Key]),
+								Dynamic: v.Dynamic,
+							}
+							if v.Permissions != "" {
+								file.Permissions = v.Permissions
+							} else if fs.ConfigMap.DefaultPermissions != "" {
+								file.Permissions = v.Permissions
```
Comment (jakefhyde) [inline pkg/provisioningv2/rke2/planner/config.go:466]: Should this be `file.Permissions = fs.ConfigMap.DefaultPermissions`?

File: pkg/provisioningv2/rke2/planner/config.go [hunk 2]
```diff
@@ -413,6 +435,72 @@ func configFile(controlPlane *rkev1.RKEControlPlane, filename string) string {
...
+					if authorized, found := clusterObjectAuthorized(secret, rke2.AuthorizedObjectAnnotation, controlPlane.Name); authorized && found {
+						for _, v := range fs.Secret.Items {
+							file := plan.File{
+								Path:    v.Path,
+								Content: base64.StdEncoding.EncodeToString(secret.Data[v.Key]),
+								Dynamic: v.Dynamic,
+							}
+							if v.Permissions != "" {
+								file.Permissions = v.Permissions
+							} else if fs.ConfigMap.DefaultPermissions != "" {
```
Comment (jakefhyde) [inline pkg/provisioningv2/rke2/planner/config.go:465]:
```suggestion
							} else if fs.Secret.DefaultPermissions != "" {
```

File: pkg/provisioningv2/rke2/planner/config.go (hunk 1 above)
Comment (jakefhyde) [inline pkg/provisioningv2/rke2/planner/config.go:466]: Should this be `file.Permissions = fs.Secret.DefaultPermissions`?

File: pkg/provisioningv2/rke2/planner/config.go [hunk 3]
```diff
@@ -413,6 +435,72 @@ func configFile(controlPlane *rkev1.RKEControlPlane, filename string) string {
...
+						for _, v := range fs.ConfigMap.Items {
+							file := plan.File{
+								Path:    v.Path,
+								Content: base64.StdEncoding.EncodeToString([]byte(configmap.Data[v.Key])),
+								Dynamic: v.Dynamic,
+							}
+							if v.Permissions != "" {
+								file.Permissions = v.Permissions
+							} else if fs.ConfigMap.DefaultPermissions != "" {
+								file.Permissions = v.Permissions
```
Comment (jakefhyde) [inline pkg/provisioningv2/rke2/planner/config.go:490]: Should this be `file.Permissions = fs.ConfigMap.DefaultPermissions`?

File: pkg/provisioningv2/rke2/planner/config.go [hunk 4]
```diff
@@ -413,6 +435,72 @@ func configFile(controlPlane *rkev1.RKEControlPlane, filename string) string {
...
 }
 
+func (p *Planner) renderFiles(controlPlane *rkev1.RKEControlPlane, entry *planEntry) ([]plan.File, error) {
+	var files []plan.File
+	for _, msf := range controlPlane.Spec.MachineSelectorFiles {
+		sel, err := metav1.LabelSelectorAsSelector(msf.MachineLabelSelector)
+		if err != nil {
+			return files, err
+		}
+		if msf.MachineLabelSelector == nil || sel.Matches(labels.Set(entry.Machine.Labels)) {
```
Comment (jakefhyde) [inline pkg/provisioningv2/rke2/planner/config.go:445]: Nit: We could invert some if statements and break this out into a function that operates on `K8sObjectFileSource` to bring the indentation down, if you're into that sort of thing. I won't prevent from merging for this though.

**Review by jakefhyde — CHANGES_REQUESTED** [review]
Couple nits, implementation looks good though

**Review by HarrisonWAffel — DISMISSED** [review]
Other than the changes requested by Jake regarding default permissions, this lgtm

File: pkg/provisioningv2/rke2/planner/config.go (hunk 1 above)
Comment (Oats87) [reply pkg/provisioningv2/rke2/planner/config.go:466]: yeah good catch

File: pkg/provisioningv2/rke2/planner/config.go (hunk 3 above)
Comment (Oats87) [reply pkg/provisioningv2/rke2/planner/config.go:490]: good catch

File: pkg/provisioningv2/rke2/planner/config.go (hunk 4 above)
Comment (Oats87) [reply pkg/provisioningv2/rke2/planner/config.go]: I'm not going to break out the function just yet, but I did invert the if statement

**Review by jakefhyde — APPROVED** [review]
LGTM

**Review by HarrisonWAffel — APPROVED** [review]

**Review by jiaqiluo — APPROVED** [review]
LGTM 💯

**Pull request merged**